keps/sig-scheduling/1923-try-nominated-node-first/README.md

```diff
@@ -0,0 +1,157 @@
+needs, the preemption could happen to make sure high priority pods could got scheduled
+and spawned.
+
+For the cluster with huge number of computing nodes, looping over all the nodes in
+the cluster for the phase of the predicate is an unnegligible consuming for scheduler,
+besides, it's unpredicatable that how the out-of-tree plugins are defined, whether
+they are time consuming, this is also a factor that may increase the time for the
+phase of predicate.
+
+### Goals
+
+- In the case of `pod.Status.NominatedNodeName` is set for the incoming pod, scheduler
+  will try the nominated node first for the phase of predicate.
+- Failed to schedule on the nominated node will fall back to verify all the nodes in
+  the cluster instead of preemption again.
+- A new feature gate will be provided for the operator to opt in/out this feature.
+
+## Proposal
+
+### User Stories (Optional)
+
+As a premium user, I want my high priority application got spin up more quickly,
+I don't care where the application is running, whether there is any other nodes
+in the cluster have more disk space, memory size etc., as long as my application
+run normally.
+
+### Notes/Constraints/Caveats (Optional)
+
+Enables this feature also implies the preemptor pod might not be dispatched
+to the best candidated node in some corner case, e.g. another node releases
+the resources and becomes the best candidate while the victim pods got removed
+from the nominated node, but enable this feature will make scheduler to consider
+the nominated node firstly.
+
+## Design Details
+
+### Implementation Details
+
+1. In the phase of predicate, which is currently implemented in the method of `findNodesThatFitPod`,
+   check the nominated directly if the incoming pod has the `pod.Status.NominatedNodeName` defined
+   and the feature gate is enabled.
+2. In case the nominated node is not suit the incoming pod anymore, return `fmt.Errorf` instead of
+   `core.FitError`, because this will give scheduler a chance to clean up the nominated node from
+   the pod and find a new node to schedule instead of preemption again (might already have during
+   the period). i.e. a fresh new scheduling cycle will be started from here.
+3. If the nominated node is not suit for the pod anymore, the scheduling failure will be recorded
+   and the `updatePod` will be called, here we change the logic to update the pod as long as the
+   parameter `nominatedNode` is different with what pod holds in `pod.Status.NominatedNodeName`.
+   In this case, parameter `nominatedNode` is an empty string so that the nominated node will be
+   cleaned from the pod and the pod will be moved to the active queue. It lets scheduler find another
+   place for the pod.
+
+
+### Test Plan
+
+Following tests wll be covered or considered:
+
+- **Unit Tests**: All core changes must be covered by unit tests.
+- **Integration Tests**: Integration test will be provided if necessary, not visible in current stage.
```

Comment:
Do a verbal description of what the test cases are supposed to look like.

We need to test for one preempted pod and several preempted pods as well.

Reply:
Yes, integration and benchmark.

Reply:
I went through the integration test, I don't think the integration test is needed, since I cannot see any difference of expected result when this feature is enabled. 
And `scheduler_perf` already has the performance testcase to cover this, which is `scheduling_algorithm_predicate_evaluation_seconds`. I am not sure more testcases are needed or not.
Anyway, I have some placeholder created in this section.